### PR TITLE
STREAM-86 Add a mechanism to choose allowed schema changes

### PR DESCRIPTION
**Pull request by akuma432**
* Implement process to choose allowed schema changes
* Added junit to test allChangesAreCompatible()
* updated templates.
* Removed copy-into from evolve-schema target

### CODE REVIEW

**Comment by akuma432** [comment]
> * Implement process to choose allowed schema changes
> * Added junit to test allChangesAreCompatible()
> * updated templates.

I will do the testing on Monday with this feature.

File: src/main/java/io/phdata/streamliner/schemadefiner/model/Jdbc.java [hunk 1]
```diff
@@ -28,6 +29,11 @@
   // for table whitelisting
   private List<String> tables;
   private int batchTableCount;
+  /* This is used to manage allowed schema changes.
+  For example: To exclude column add for CSV users we should not pass "COLUMN_ADD" in the list. Then column add becomes incompatible change for CSV users.
+  If this parameter is not provided in config, by default it includes all the values from enum SchemaChanges. This is handled in code.
+  */
+  private List<SchemaChanges> validSchemaChanges = new ArrayList<>();
```
Comment (brockn) [inline src/main/java/io/phdata/streamliner/schemadefiner/model/Jdbc.java:36]: Should this be a Set?

File: src/main/java/io/phdata/streamliner/schemadefiner/model/SchemaChanges.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package io.phdata.streamliner.schemadefiner.model;
+
+public enum SchemaChanges {
+  /* Currently schema evolution feature supported by Streamliner 5.x */
+  TABLE_ADD("new-table-add"),
```
[the remark this answers is not in the document]
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/model/SchemaChanges.java]: updated

File: src/main/java/io/phdata/streamliner/schemadefiner/model/SchemaChanges.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  COLUMN_ADD("new-column-add"),
+  UPDATE_COLUMN_COMMENT("update-column-comment"),
```
[the remark this answers is not in the document]
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/model/SchemaChanges.java]: updated

File: src/main/java/io/phdata/streamliner/schemadefiner/model/Jdbc.java (hunk 1 above)
Comment (akuma432) [reply src/main/java/io/phdata/streamliner/schemadefiner/model/Jdbc.java]: updated

**Comment by akuma432** [comment]
> * Implement process to choose allowed schema changes
> * Added junit to test allChangesAreCompatible()
> * updated templates.
> * Removed copy-into from evolve-schema target

Included this change also. https://phdata.atlassian.net/browse/STREAM-87

**Comment by akuma432** [comment]
> > * Implement process to choose allowed schema changes
> > * Added junit to test allChangesAreCompatible()
> > * updated templates.
> 
> I will do the testing on Monday with this feature.

Tested few use cases with example templates and MDT templates, it looks good. 
One thing is currently we only implemented allowed schema changes for an existing table only. Meaning we can control column changes in an existing table.
In case a new table is created it is considered as allowed change by default. Please let me know if we want to control this as well. If yes then I will raise a separate PR.

**Comment by akuma432** [comment]
> One thing is currently we only implemented allowed schema changes for an existing table only. Meaning we can control column changes in an existing table.
> In case a new table is created it is considered as allowed change by default. Please let me know if we want to control this as well. If yes then I will raise a separate PR.

Creation of new table is also handled now in this PR only.

**Review by brockn — APPROVED** [review]

**Pull request merged**